File: DSA/DSA_450/Python/02_matrix/001_Spiral_Traversal_Matrix.py
"""
Problem: Spiral Traversal of a Matrix
URL: https://practice.geeksforgeeks.org/problems/spirally-traversing-a-matrix-1587115621/1

Problem Statement:
Given a matrix of size R x C, print the elements in spiral order traversal.

Sample Input/Output:
Input: matrix = [[1, 2, 3, 4],
                 [5, 6, 7, 8],
                 [9, 10, 11, 12],
                 [13, 14, 15, 16]]
Output: [1, 2, 3, 4, 8, 12, 16, 15, 14, 13, 9, 5, 6, 7, 11, 10]

Input: matrix = [[1, 2, 3],
                 [4, 5, 6],
                 [7, 8, 9]]
Output: [1, 2, 3, 6, 9, 8, 7, 4, 5]
"""


class Solution:
    def Spiral_Traversal_Boundary_Optimal(self, matrix):
        """
        Boundary Shrinking - Track top, bottom, left, right boundaries
        Time Complexity: O(m * n)
        Space Complexity: O(1) excluding result
        """
        ans = []
        if not matrix:
            return ans
        top, bottom, right, left = 0, len(matrix), len(matrix[0]), 0
        while top < bottom and left < right:
            for i in range(left, right):
                ans.append(matrix[top][i])
            top += 1
            for i in range(top, bottom):
                ans.append(matrix[i][right - 1])
            right -= 1
            if top < bottom:
                for i in range(right - 1, left - 1, -1):
                    ans.append(matrix[bottom - 1][i])
                bottom -= 1
            if left < right:
                for i in range(bottom - 1, top - 1, -1):
                    ans.append(matrix[i][left])
                left += 1
        return ans

    def Spiral_Traversal_Direction_Array(self, matrix):
        """
        Direction Array - Use direction vectors and visited tracking
        Time Complexity: O(m * n)
        Space Complexity: O(m * n)
        """
        ans = []
        if not matrix:
            return ans
        R, C = len(matrix), len(matrix[0])
        seen = [[False] * C for _ in range(R)]
        dr = [0, 1, 0, -1]
        dc = [1, 0, -1, 0]
        r, c, di = 0, 0, 0
        for _ in range(R * C):
            ans.append(matrix[r][c])
            seen[r][c] = True
            cr, cc = r + dr[di], c + dc[di]
            if 0 <= cr < R and 0 <= cc < C and not seen[cr][cc]:
                r, c = cr, cc
            else:
                di = (di + 1) % 4
                r += dr[di]
                c += dc[di]
        return ans


def Test_Spiral_Traversal():
    solution = Solution()

    test_cases = [
        [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12], [13, 14, 15, 16]],
        [[1, 2, 3], [4, 5, 6], [7, 8, 9]],
        [[1, 2, 3, 4], [5, 6, 7, 8]],
        [[1], [2], [3]]
    ]

    for matrix in test_cases:
        print("Matrix:")
        for row in matrix:
            print("\t".join(str(x) for x in row))

        r1 = solution.Spiral_Traversal_Boundary_Optimal(matrix)
        print("Boundary:", " ".join(str(x) for x in r1))

        r2 = solution.Spiral_Traversal_Direction_Array(matrix)
        print("Direction:", " ".join(str(x) for x in r2))

        print("-" * 50)


if __name__ == "__main__":
    Test_Spiral_Traversal()

File: DSA/DSA_450/Cpp/02_matrix/001_Spiral_Traversal_Matrix.cpp
/*
Problem: Spiral Traversal of a Matrix
URL: https://practice.geeksforgeeks.org/problems/spirally-traversing-a-matrix-1587115621/1

Problem Statement:
Given a matrix of size R x C, print the elements in spiral order traversal.

Sample Input/Output:
Input: matrix = [[1, 2, 3, 4],
                 [5, 6, 7, 8],
                 [9, 10, 11, 12],
                 [13, 14, 15, 16]]
Output: [1, 2, 3, 4, 8, 12, 16, 15, 14, 13, 9, 5, 6, 7, 11, 10]

Input: matrix = [[1, 2, 3],
                 [4, 5, 6],
                 [7, 8, 9]]
Output: [1, 2, 3, 6, 9, 8, 7, 4, 5]
*/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> Spiral_Traversal_Boundary_Optimal(vector<vector<int>>& matrix) {
        /*
        Boundary Shrinking - Track top, bottom, left, right boundaries
        Time Complexity: O(m * n)
        Space Complexity: O(1) excluding result
        */
        vector<int> ans;
        if (matrix.empty()) return ans;
        int top = 0, bottom = matrix.size(), right = matrix[0].size(), left = 0;
        while (top < bottom && left < right) {
            for (int i = left; i < right; i++) ans.push_back(matrix[top][i]);
            top++;
            for (int i = top; i < bottom; i++) ans.push_back(matrix[i][right - 1]);
            right--;
            if (top < bottom) {
                for (int i = right - 1; i >= left; i--) ans.push_back(matrix[bottom - 1][i]);
                bottom--;
            }
            if (left < right) {
                for (int i = bottom - 1; i >= top; i--) ans.push_back(matrix[i][left]);
                left++;
            }
        }
        return ans;
    }

    vector<int> Spiral_Traversal_Direction_Array(vector<vector<int>>& matrix) {
        /*
        Direction Array - Use direction vectors and visited tracking
        Time Complexity: O(m * n)
        Space Complexity: O(m * n)
        */
        vector<int> ans;
        if (matrix.empty()) return ans;
        int R = matrix.size(), C = matrix[0].size();
        vector<vector<bool>> seen(R, vector<bool>(C, false));
        int dr[] = {0, 1, 0, -1};
        int dc[] = {1, 0, -1, 0};
        int r = 0, c = 0, di = 0;
        for (int i = 0; i < R * C; i++) {
            ans.push_back(matrix[r][c]);
            seen[r][c] = true;
            int cr = r + dr[di], cc = c + dc[di];
            if (cr >= 0 && cr < R && cc >= 0 && cc < C && !seen[cr][cc]) {
                r = cr;
                c = cc;
            } else {
                di = (di + 1) % 4;
                r += dr[di];
                c += dc[di];
            }
        }
        return ans;
    }
};

void Test_Spiral_Traversal() {
    Solution solution;

    vector<vector<vector<int>>> test_cases = {
        {{1,2,3,4},{5,6,7,8},{9,10,11,12},{13,14,15,16}},
        {{1,2,3},{4,5,6},{7,8,9}},
        {{1,2,3,4},{5,6,7,8}},
        {{1},{2},{3}}
    };

    for (auto& matrix : test_cases) {
        cout << "Matrix:" << endl;
        for (auto& row : matrix) {
            for (int x : row) cout << x << "\t";
            cout << endl;
        }

        auto r1 = solution.Spiral_Traversal_Boundary_Optimal(matrix);
        cout << "Boundary: ";
        for (int x : r1) cout << x << " ";
        cout << endl;

        auto r2 = solution.Spiral_Traversal_Direction_Array(matrix);
        cout << "Direction: ";
        for (int x : r2) cout << x << " ";
        cout << endl;

        cout << string(50, '-') << endl;
    }
}

int main() {
    Test_Spiral_Traversal();
    return 0;
}
